### ISSUE TITLE
💅 False positive for `noUnusedImports` when using `import * as name` in a type-only position

### ISSUE THREAD

**Issue by flying-sheep** [biomejs/biome#6835]
### Environment information

```bash
CLI:
  Version:                      2.1.1
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  BIOME_THREADS:                unset
  NO_COLOR:                     unset
  TERM:                         xterm-256color
  JS_RUNTIME_VERSION:           v20.2.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         yarn/4.5.3

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         biome.jsonc
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  false

Linter:
  JavaScript enabled:           unset
  JSON enabled:                 unset
  CSS enabled:                  unset
  GraphQL enabled:              unset
  Recommended:                  unset
  Enabled rules:
    a11y/noAccessKey
    a11y/noAriaHiddenOnFocusable
    a11y/noAriaUnsupportedElements
    a11y/noAutofocus
    a11y/noDistractingElements
    a11y/noHeaderScope
    a11y/noInteractiveElementToNoninteractiveRole
    a11y/noLabelWithoutControl
    a11y/noNoninteractiveElementToInteractiveRole
    a11y/noNoninteractiveTabindex
    a11y/noPositiveTabindex
    a11y/noRedundantAlt
    a11y/noRedundantRoles
    a11y/noStaticElementInteractions
    a11y/noSvgWithoutTitle
    a11y/useAltText
    a11y/useAnchorContent
    a11y/useAriaActivedescendantWithTabindex
    a11y/useAriaPropsForRole
    a11y/useAriaPropsSupportedByRole
    a11y/useButtonType
    a11y/useFocusableInteractive
    a11y/useGenericFontNames
    a11y/useHeadingContent
    a11y/useHtmlLang
    a11y/useIframeTitle
    a11y/useKeyWithClickEvents
    a11y/useKeyWithMouseEvents
    a11y/useMediaCaption
    a11y/useSemanticElements
    a11y/useValidAnchor
    a11y/useValidAriaProps
    a11y/useValidAriaRole
    a11y/useValidAriaValues
    a11y/useValidAutocomplete
    a11y/useValidLang
    complexity/noAdjacentSpacesInRegex
    complexity/noArguments
    complexity/noBannedTypes
    complexity/noCommaOperator
    complexity/noEmptyTypeParameters
    complexity/noExtraBooleanCast
    complexity/noFlatMapIdentity
    complexity/noStaticOnlyClass
    complexity/noThisInStatic
    complexity/noUselessCatch
    complexity/noUselessConstructor
    complexity/noUselessContinue
    complexity/noUselessEmptyExport
    complexity/noUselessEscapeInRegex
    complexity/noUselessFragments
    complexity/noUselessLabel
    complexity/noUselessLoneBlockStatements
    complexity/noUselessRename
    complexity/noUselessStringRaw
    complexity/noUselessSwitchCase
    complexity/noUselessTernary
    complexity/noUselessThisAlias
    complexity/noUselessTypeConstraint
    complexity/noUselessUndefinedInitialization
    complexity/useArrowFunction
    complexity/useDateNow
    complexity/useFlatMap
    complexity/useLiteralKeys
    complexity/useNumericLiterals
    complexity/useOptionalChain
    complexity/useRegexLiterals
    complexity/useSimpleNumberKeys
    correctness/noChildrenProp
    correctness/noConstAssign
    correctness/noConstantCondition
    correctness/noConstantMathMinMaxClamp
    correctness/noConstructorReturn
    correctness/noEmptyCharacterClassInRegex
    correctness/noEmptyPattern
    correctness/noGlobalObjectCalls
    correctness/noInnerDeclarations
    correctness/noInvalidBuiltinInstantiation
    correctness/noInvalidConstructorSuper
    correctness/noInvalidDirectionInLinearGradient
    correctness/noInvalidGridAreas
    correctness/noInvalidPositionAtImportRule
    correctness/noInvalidUseBeforeDeclaration
    correctness/noMissingVarFunction
    correctness/noNonoctalDecimalEscape
    correctness/noPrecisionLoss
    correctness/noSelfAssign
    correctness/noSetterReturn
    correctness/noStringCaseMismatch
    correctness/noSwitchDeclarations
    correctness/noUnknownFunction
    correctness/noUnknownMediaFeatureName
    correctness/noUnknownProperty
    correctness/noUnknownPseudoClass
    correctness/noUnknownPseudoElement
    correctness/noUnknownTypeSelector
    correctness/noUnknownUnit
    correctness/noUnmatchableAnbSelector
    correctness/noUnreachable
    correctness/noUnreachableSuper
    correctness/noUnsafeFinally
    correctness/noUnsafeOptionalChaining
    correctness/noUnusedFunctionParameters
    correctness/noUnusedImports
    correctness/noUnusedLabels
    correctness/noUnusedPrivateClassMembers
    correctness/noUnusedVariables
    correctness/noVoidElementsWithChildren
    correctness/noVoidTypeReturn
    correctness/useIsNan
    correctness/useValidForDirection
    correctness/useValidTypeof
    correctness/useYield
    performance/noAccumulatingSpread
    performance/noDynamicNamespaceImportAccess
    security/noBlankTarget
    security/noDangerouslySetInnerHtml
    security/noDangerouslySetInnerHtmlWithChildren
    security/noGlobalEval
    style/noDescendingSpecificity
    style/noNonNullAssertion
    style/useArrayLiterals
    style/useConst
    style/useDeprecatedReason
    style/useExponentiationOperator
    style/useExportType
    style/useImportType
    style/useLiteralEnumMembers
    style/useNodejsImportProtocol
    style/useShorthandFunctionType
    style/useTemplate
    suspicious/noApproximativeNumericConstant
    suspicious/noArrayIndexKey
    suspicious/noAssignInExpressions
    suspicious/noAsyncPromiseExecutor
    suspicious/noCatchAssign
    suspicious/noClassAssign
    suspicious/noCommentText
    suspicious/noCompareNegZero
    suspicious/noConfusingLabels
    suspicious/noConfusingVoidType
    suspicious/noConstEnum
    suspicious/noControlCharactersInRegex
    suspicious/noDebugger
    suspicious/noDocumentCookie
    suspicious/noDoubleEquals
    suspicious/noDuplicateAtImportRules
    suspicious/noDuplicateCase
    suspicious/noDuplicateClassMembers
    suspicious/noDuplicateCustomProperties
    suspicious/noDuplicateElseIf
    suspicious/noDuplicateFields
    suspicious/noDuplicateFontNames
    suspicious/noDuplicateJsxProps
    suspicious/noDuplicateObjectKeys
    suspicious/noDuplicateParameters
    suspicious/noDuplicateProperties
    suspicious/noDuplicateSelectorsKeyframeBlock
    suspicious/noEmptyBlock
    suspicious/noEmptyInterface
    suspicious/noExplicitAny
    suspicious/noExtraNonNullAssertion
    suspicious/noFallthroughSwitchClause
    suspicious/noFunctionAssign
    suspicious/noGlobalAssign
    suspicious/noGlobalIsFinite
    suspicious/noGlobalIsNan
    suspicious/noImplicitAnyLet
    suspicious/noImportAssign
    suspicious/noImportantInKeyframe
    suspicious/noIrregularWhitespace
    suspicious/noLabelVar
    suspicious/noMisleadingCharacterClass
    suspicious/noMisleadingInstantiator
    suspicious/noMisrefactoredShorthandAssign
    suspicious/noOctalEscape
    suspicious/noPrototypeBuiltins
    suspicious/noRedeclare
    suspicious/noRedundantUseStrict
    suspicious/noSelfCompare
    suspicious/noShadowRestrictedNames
    suspicious/noShorthandPropertyOverrides
    suspicious/noSparseArray
    suspicious/noSuspiciousSemicolonInJsx
    suspicious/noTemplateCurlyInString
    suspicious/noThenProperty
    suspicious/noUnsafeDeclarationMerging
    suspicious/noUnsafeNegation
    suspicious/noWith
    suspicious/useAdjacentOverloadSignatures
    suspicious/useDefaultSwitchClauseLast
    suspicious/useGetterReturn
    suspicious/useGoogleFontDisplay
    suspicious/useIsArray
    suspicious/useNamespaceKeyword

Workspace:
  Open Documents:               0
```

### Rule name

noUnusedImports

### Playground link

https://biomejs.dev/playground/?code=aQBtAHAAbwByAHQAIAAqACAAYQBzACAAcABvAHMAdABjAHMAcwBNAG8AZAB1AGwAZQBzACAAZgByAG8AbQAgACIAcABvAHMAdABjAHMAcwAtAG0AbwBkAHUAbABlAHMAIgAKAAoAdAB5AHAAZQAgAFAAbwBzAHQAYwBzAHMATwBwAHQAaQBvAG4AcwAgAD0AIABQAGEAcgBhAG0AZQB0AGUAcgBzADwAcABvAHMAdABjAHMAcwBNAG8AZAB1AGwAZQBzAD4AWwAwAF0ACgAKAGUAeABwAG8AcgB0ACAAZgB1AG4AYwB0AGkAbwBuACAAZgAoAG8AcAB0AGkAbwBuAHMAOgAgAFAAbwBzAHQAYwBzAHMATwBwAHQAaQBvAG4AcwApACAAewAKACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKABvAHAAdABpAG8AbgBzACkACgB9AAoA

### Expected result

Biome should realize it’s used.

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

**Comment by siketyan** [comment]
I don't think this is a valid example, as namespace import will never be a type in ESM. Did you mean using default import instead?

**Comment by ematipico** [comment]
I believe this is an incorrect case. I think  it's not possible to use a namespace import, and use it as a type. 

With namespace imports, we're essentially importing everything that a module is exporting, which means a correct usage would be something like `postcssModules.TypeToUse`

**Comment by arendjr** [comment]
I don't understand the confusion here? The example in the playground link is a valid use case. It doesn't matter that the namespace import isn't _just_ a type, it can still be used in a type position. You can copy the same example into the TypeScript playground and it will just work: https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM50qmBjVqCyEAJgK4A2ApugGZQQhwBEmOeAtCMeVYwFC8wAnmApwAChCy5UAeTAxgEAHboAvOORRkICjApRUAHhbTCpSqgB8AbQAMAXX4UAHpFhxqJJdgXKPACgh5RRUALnFJVllg5VQASjgAb144OGxYiEoAOjIIAHNAmJU43gBfXiA

**Comment by flying-sheep** [comment]
@siketyan and @ematipico are probably confused because they don’t know about [`export =`](https://www.typescriptlang.org/docs/handbook/modules/reference.html#export--and-import--require) syntax.

But I agree: a good response to being confused is to just try things out (as opposed to assuming that they’re invalid). Especially if someone else already triaged the issue and marked it as valid.

**Comment by ematipico** [comment]
It's also OK to be wrong :) please be kind and show some patience

**Comment by flying-sheep** [comment]
I’m wrong a lot! I’m just sharing my experience as a maintainer about the strategy I’ve adopted to deal with that.

**Comment by siketyan** [comment]
In my understanding the `export =` syntax is only for CJS compatibility and is invalid in ESM. I understand many people still use CJS, but isn't it also true we can use default import to use the stuff exported by `export =`?

**Comment by flying-sheep** [comment]
> In my understanding the export = syntax is only for CJS compatibility and is invalid in ESM.

That’s true in the context of ECMAScript’s ESM support. The syntax *is* valid TypeScript though, so a project claiming to support TypeScript needs to support it.

> but isn't it also true we can use default import to use the stuff exported by `export =`?

That’s not true, you import modules defined by `export =` using `import * as ... from 'module-name'`, like I did in my playground example.

**Comment by flying-sheep** [comment]
Also let’s not get distracted: even if the syntax didn’t exist, using the the variable created by `import * as name from '...'` in a type expression is valid in any case.

**Comment by flying-sheep** [comment]
This is still the case

**Comment by Conaclos** [comment]
> even if the syntax didn’t exist, using the the variable created by import * as name from '...' in a type expression is valid in any case.

To my knowledge, that's not true. You got the following error: `Cannot use namespace 'X' as a type.`

I thought that `export =` was obsolete syntax for CJS compat?

**Comment by flying-sheep** [comment]
If it’s obsolete syntax, what’s the replacement syntax for CJS compat?

Or do you mean that CJS compat is obsolete? Sadly that’s not the case, there are still a bunch of important, (otherwise) well-maintained packages that only ship CJS.
